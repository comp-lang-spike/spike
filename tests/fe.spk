#!/usr/bin/env spike

printf(format, ...args) { stdout.printf(format, ...args); }
puts(s) { stdout.printf("%s\n", s); }

abc() {
    obj alphabet, letter, i;
    
    alphabet = "abcdefghijklmnopqrstuvwxyz";
    for (i = 0; i < alphabet.len; ++i) {
        letter = alphabet[i];
        printf("%c ", letter);
    }
    puts("");
    
    obj aBlock;
    aBlock = [ :a :b :c |
              printf("little girl voice: %c %c %c\n", a, b, c); ];
    
    aBlock('a', 'b', 'c');
    obj letters;
    letters = tuple('x', 'y', 'z');
    aBlock(...letters);
    letters = tuple('j', 'k');
    aBlock('i', ...letters);
    
    alphabet do: [ :l | printf("%c ", l); ];
    puts("");
    
    alphabet do: [ :l |
        printf("%c ", l);
        ];
    puts("");
    
    varargs("Now", "I", "know", "my", "ABCs.");
    
    alphabet do: [ :l | printf("%c ", l); if (l == 'g') return "hey"; ];
    puts("ERROR: not reached");
}

tuple(...args) {
    return args;
}

varargs(...args) {
    args do: [ :a | printf("%s\n", a); ];
}

main(argv) {
    puts(abc());
    return 0;
}
