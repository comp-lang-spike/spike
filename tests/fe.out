a b c d e f g h i j k l m n o p q r s t u v w x y z 
little girl voice: a b c
little girl voice: x y z
little girl voice: i j k
a b c d e f g h i j k l m n o p q r s t u v w x y z 
a b c d e f g h i j k l m n o p q r s t u v w x y z 
Now
I
know
my
ABCs.
a b c d e f g hey
